edgeml_pytorch/trainer/bonsaiTrainer.py

```diff
@@ -126,13 +126,13 @@ def runHardThrsd(self):
             __thrsdT).to(self.device)
 
         self.__thrsdW = torch.FloatTensor(
-            np.copy(__thrsdW)).to(self.device)
```

Comment:
@adityakusupati  Are you OK with this change?

Reply:
@harsha-simhadri , this was the issue I mention about Bonsai when we changed HardThreshold function. 
@pushkalkatara , thanks a lot for taking this initiative (I was supposed to do this) can you please check the flow of gradients? I am pretty sure clone() allows gradient propagation and we don't want out intermediate state variables to change due to SGD.  https://pytorch.org/docs/stable/tensors.html#torch.Tensor.clone. I will approve this only when self.__thrsdW has no links to __thrsdW in the computational graph. __thrsdW's memory location is shared by self.Bonsai.W which will create problems. There should be a detach somewhere and please print the tensor values during optimization to be sure.

I will set the repo up in next couple of days hopefully and will help you out.

Reply:
Maybe we can use
```python
self.__thrsdW = torch.FloatTensor(
            __thrsdW.clone().detach().to(self.device)
```
I believe it would make a copy of __thrsdW(with data and reference in the computational graph) and further, .detach() would remove the link too and disable differentiation.

Reply:
@adityakusupati  Any thoughts?

Reply:
The reply went to the bottom of the thread.

Technically this should work. However I had some bad experience using
clone.detach where gradient did flow through. Can you please check/debug
before you commit. @pushkalkatara 

Aditya

Reply:
@adityakusupati it seems to be training for me with this @pushkalkatara  change, can you show me a script that failed to run

Reply:
@harsha-simhadri I never said that this won't run. But the internals of training might vary. The code will run without any issues. I am worried about the propagation of gradient during the non-iht steps as there is some sort of memory sharing that is happening. All we need to do is for a given run where, once the iht happens, we need to ensure that base tensor isn't changing due to back prop.

In short, yes this fix will work, but this needs testing in regards to state change due to backprop due to my doubts/previous hardships on clone().detach()

Reply:
@harsha-simhadri and @pushkalkatara this - https://github.com/microsoft/EdgeML/blob/harsha/reorg/edgeml_pytorch/graph/rnn.py#L152
 will work. This is the safest fix. clone().detach() did fail in this case and that is the reason we moved to the current implementation

Reply:
can you propose any concrete tests? or do you want me to go through the entire code and  reason where this could go wrong?

Reply:
@harsha-simhadri let me try to come up with the tests. But if @pushkalkatara's fix works then there wouldn't have been a need for [this](https://github.com/microsoft/EdgeML/blob/harsha/reorg/edgeml_pytorch/graph/rnn.py#L152).  I will get back to you. It is not our code that is a problem, it is PyTorch's issue.

Reply:
@pushkalkatara and @harsha-simhadri I have tested this change extensively on a cpu and there is no difference. But given the documentation and my understanding I request the change to be
`self.__thrsdW = torch.FloatTensor(
            __thrsdW.detach().clone().to(self.device))`
This will ensure that the initial tensor is decoupled with graph and has no gradient followed by the making of a memory copy.
@SachinG007  please test this on a gpu and see the final model dump and ensure that the sparsity is intact.

I was very prejudiced by my previous experience and the fact we still have [this](https://github.com/microsoft/EdgeML/blob/harsha/reorg/edgeml_pytorch/graph/rnn.py#L152). I am sorry.

Reply:
I'll commit the changes. Should I make the changes [here](https://github.com/microsoft/EdgeML/blob/harsha/reorg/edgeml_pytorch/graph/rnn.py#L152) also?

Reply:
yes please do, I made these changes locally, and bonsai and fastgrnn seem fine on cpu and gpu. We will revert later if something bad happens. @adityakusupati is that alright?

Reply:
@harsha-simhadri and @pushkalkatara  yes, I am fine with it.